### PR TITLE
WFLY-4780 Add migrate operation to the web subsystem to allow it to b…

### PR DESCRIPTION
**Pull request by stuartwdouglas**
…e migrated to the Undertow subsystem

### CODE REVIEW

**Comment by stuartwdouglas** [comment]
Retest this please

File: legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+        }
+        if(certificateFile.isDefined()) {
+            UndertowLogger.ROOT_LOGGER.couldNotMigrateResource(WebSSLDefinition.CERTIFICATE_FILE.getName(), pathAddress(legacyAddOp.get(ADDRESS)));
+        }
+
+        if(csRevocationURL.isDefined()) {
+            UndertowLogger.ROOT_LOGGER.couldNotMigrateResource(WebSSLDefinition.CA_REVOCATION_URL.getName(), pathAddress(legacyAddOp.get(ADDRESS)));
+        }
+        if(sslProtocol.isDefined()) {
+            UndertowLogger.ROOT_LOGGER.couldNotMigrateResource(WebSSLDefinition.SSL_PROTOCOL.getName(), pathAddress(legacyAddOp.get(ADDRESS)));
```
Comment (ctomc) [inline legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java]: this is available on https-listener attribute enabled-protocols

**Comment by stuartwdouglas** [comment]
@bstansberry does this look ok to you?

File: legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java [hunk 2]
```diff
@@ -0,0 +1,725 @@
...
+
+                createWelcomeContentHandler(migrationOperations);
+
+                // transform the legacy add operations and put them in migrationOperations
+                transformResources(context, legacyModelAddOps, migrationOperations);
+
+                fixAddressesForDomainMode(pathAddress(operation.get(ADDRESS)), migrationOperations);
+
+                // put the /subsystem=messaging:remove operation
+                removeWebSubsystem(migrationOperations);
```
Comment (jmesnil) [inline legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java]: this should be done before `fixAddressesForDomainMode`

File: legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java [hunk 3]
```diff
@@ -0,0 +1,725 @@
...
+    public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+        if (!describe && context.getRunningMode() != RunningMode.ADMIN_ONLY) {
+            throw UndertowLogger.ROOT_LOGGER.migrateOperationAllowedOnlyInAdminOnly();
+        }
+
+        // node containing the description (list of add operations) of the legacy subsystem
+        final ModelNode legacyModelAddOps = new ModelNode();
+        //we don't preserve order, instead we sort by address length
+        //TODO: is this ok in every case?
+        final Map<PathAddress, ModelNode> migrationOperations = new TreeMap<>(new Comparator<PathAddress>() {
```
Comment (jmesnil) [inline legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java]: fwiw, for the messaging subsystem, I used a LinkedHashMap to rely on the order of the operations returned by the :describe operations.

**Comment by bstansberry** [comment]
This shouldn't block this, but we should look into it for web and messaging.

These subsystems should not be usable on a non-admin-only server. Their function is to provide a model for migration, but if a runtime is being set up they should fail.

LegacyExtension would handle that for something like cmp, but it can't be used here of course. But these subsystems are still using ModelOnlyAddStepHandler/ModelOnlyWriteAttributeHandler/ModelOnlyRemoveStepHandler and those handlers will happily work in a non-admin-only server, just storing the model and doing nothing in runtime. For these subsystems we should fail in performRuntime if it's called.

**Comment by bstansberry** [comment]
I'll try and look more later.

**Comment by bstansberry** [comment]
A simple solution to my previous long comment would be for LegacyExtension to support a boolean constructor param. If 'false' (default) the behavior is like now. If true it doesn't complain about being run on a standalone server if ExtensionContext.getRunningMode() == RunningMode.ADMIN_ONLY.

File: legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+        add.get(Constants.PATH).set(new ModelNode(new ValueExpression("${jboss.home.dir}/welcome-content")));
+        migrationOperations.put(address, add);
+    }
+
+    private ModelNode createAddOp(PathAddress address) {
+        ModelNode add = new ModelNode();
+        add.get(OP).set(ADD);
+        add.get(ADDRESS).set(address.toModelNode());
+        return add;
+    }
```
Comment (bstansberry) [inline legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java]: Not a problem, but there's org.jboss.as.controller.operations.common.Util.createAddOperation(PathAddress address)

Comment (bstansberry) [inline legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java]: which you import. ;)

**Comment by bstansberry** [comment]
@stuartwdouglas other than Jeff's comment, looks good.

File: legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java (hunk 3 above)
Comment (stuartwdouglas) [inline legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java]: Undertow does not have the same 1:1 mapping that messaging has, which was why I changed to a sorted map.

**Pull request merged**